src/topology/noetherian_space.lean

```diff
@@ -0,0 +1,217 @@
+/-
+Copyright (c) 2022 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import topology.sets.compacts
+import order.order_iso_nat
+import order.compactly_generated
+
+/-!
+# Noetherian space
+
+A Noetherian space is a topological space that satisfies any of the following equivalent conditions:
+- `well_founded ((>) : opens α → opens α → Prop)`
+- `well_founded ((<) : closeds α → closeds α → Prop)`
+- `∀ s : set α, is_compact s`
+- `∀ s : opens α, is_compact s`
+
+The first is chosen as the definition, and the equivalence is shown in
+`topological_space.noetherian_space_tfae`.
+
+Many examples of noetherian spaces come from algebraic topology. For example, the underlying space
+of a noetherian scheme (and spectrums of noetherian rings in particular) is noetherian.
+
+## Main Results
+- `noetherian_space.set`: Every subspace of a noetherian space is noetherian.
+- `noetherian_space.is_compact`: Every subspace of a noetherian space is compact.
+- `noetherian_space_tfae`: Describes the equivalent definitions of noetherian spaces.
+- `noetherian_space.range`: The image of a noetherian space under a continuous map is noetherian.
+- `noetherian_space.Union`: The finite union of noetherian spaces is noetherian.
+- `noetherian_space.discrete`: A noetherian and hausdorff space is discrete.
+- `noetherian_space.exists_finset_irreducible` : Every closed subset of a noetherian space is a
+  finite union of irreducible closed subsets.
+- `noetherian_space.finite_irreducible_components `: The number of irreducible components of a
+  noetherian space is finite.
+
+-/
+
+variables (α β : Type*) [topological_space α] [topological_space β]
+
+namespace topological_space
+
+/-- Type class for noetherian spaces. It is defined to be spaces whose open sets satisfies ACC. -/
+@[mk_iff]
+class noetherian_space : Prop :=
+(well_founded : well_founded ((>) : opens α → opens α → Prop))
```

Comment:
Just a small note: if all goes well, #15023 will change this predicate to `well_founded_lt (opens α)`.

Reply:
In fact, if that PR is merged, what should be done is change `noetherian_space` to a `reducible def`. That way, all pre-existing API for `well_founded_gt` will work seamlessly.

Reply:
If you think this PR is close to being merged, then I'll add it as a dependency to my own, and I'll make that change. My refactor will take quite a while...